letsencrypt/client.py

```diff
@@ -383,57 +383,82 @@ def deploy_certificate(self, domains, privkey_path,
+            :meth:`argparse.ArgumentParser.parse_args`.
+            it must have the redirect, hsts and uir attributes.
+        :type namespace: :class:`argparse.Namespace`
 
         :raises .errors.Error: if no installer is specified in the
             client.
 
         """
+
         if self.installer is None:
             logger.warning("No installer is specified, there isn't any "
                            "configuration to enhance.")
             raise errors.Error("No installer available")
 
+        if config is None:
+            logger.warning("No config is specified.")
+            raise errors.Error("No config available")
+
+        redirect = config.redirect
+        hsts = config.hsts
+        uir = config.uir # Upgrade Insecure Requests
+
         if redirect is None:
             redirect = enhancements.ask("redirect")
 
-        # When support for more enhancements are added, the call to the
-        # plugin's `enhance` function should be wrapped by an ErrorHandler
         if redirect:
-            self.redirect_to_ssl(domains)
+            self.apply_enhancement(domains, "redirect")
+
+        if hsts:
+            self.apply_enhancement(domains, "http-header",
+                    "Strict-Transport-Security")
+        if uir:
+            self.apply_enhancement(domains, "http-header",
+                    "Upgrade-Insecure-Requests")
+
+        msg = ("We were unable to restart web server")
+        if redirect or hsts or uir:
+            with error_handler.ErrorHandler(self._rollback_and_restart, msg):
+                self.installer.restart()
+
+    def apply_enhancement(self, domains, enhancement, options=None):
+        """Applies an enhacement on all domains.
 
-    def redirect_to_ssl(self, domains):
-        """Redirect all traffic from HTTP to HTTPS
+        :param domains: list of ssl_vhosts
+        :type list of str
 
-        :param vhost: list of ssl_vhosts
-        :type vhost: :class:`letsencrypt.interfaces.IInstaller`
+        :param enhancement: name of enhancement, e.g. http-header
+        :type str
+
+        .. note:: when more options are need make options a list.
+        :param options: options to enhancement, e.g. Strict-Transport-Security
+        :type str
+
+        :raises .errors.Error: if no installer is specified in the
+            client.
 
         """
-        msg = ("We were unable to set up a redirect for your server, "
-               "however, we successfully installed your certificate.")
+        msg = ("We were unable to set up enhancement %s for your server, "
+               "however, we successfully installed your certificate."
+               % (enhancement))
         with error_handler.ErrorHandler(self._recovery_routine_with_msg, msg):
             for dom in domains:
                 try:
-                    self.installer.enhance(dom, "redirect")
+                    self.installer.enhance(dom, enhancement, options)
                 except errors.PluginError:
```

Comment:
Per IRC:
```
<sagi__> pdeee: I'll change it to ensure-http-header. Regarding the not crashing if enhancements are already in place -
<sagi__> It's not that easy
<pdeee> not least, because of bmw's great error handling code :)
<pdeee> for your new header calls, I think we know that no recovery routine is required
<sagi__> yes, the problem is with the redirect
<pdeee> we need to check in the case of redirect...
<pdeee> so if redirect needs real recovery, what we'd do is:
<pdeee> move the "with error_handler.ErrorHandler" logic inside of the for loop
<pdeee> create a new error handler type, which knows about these non-fatal enhancement errors
<pdeee> and cleans up if required but doesn't exit out when they happen
```